### PR TITLE
support hugegraph on tikv 1st version

### PR DESCRIPTION
**Pull request by zhoney**

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+            E.checkArgument(args.length == 1,
+                            "The args count of %s must be 1", meta);
+            long splitSize = (long) args[0];
+            return this.shardSpliter.getSplits(session, splitSize);
+        });
+    }
+
+    @Override
+    public void init(Session session) {
+        // pass
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: add TODO?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    @Override
+    public void clear(Session session) {
+        // pass
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: ditto

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvSessions.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    public abstract void createTable(String... tables);
+    public abstract void dropTable(String... tables);
+    public abstract boolean existsTable(String table);
+
+    @Override
+    public abstract Session session();
+
+    /**
+     * Session for RocksDB
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvSessions.java]: improve comments  related to "RocksDB"

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+        // Query by range
+        if (query instanceof IdRangeQuery) {
+            IdRangeQuery rq = (IdRangeQuery) query;
+            return this.queryByRange(session, rq);
+        }
+
+        // Query by id
+        if (query.conditions().isEmpty()) {
+            assert !query.ids().isEmpty();
+            // NOTE: this will lead to lazy create rocksdb iterator
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: ditto

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+        public boolean hasChanges() {
+            return this.size() > 0;
+        }
+
+        /**
+         * Commit all updates(put/delete) to DB
+         */
+        @Override
+        public Integer commit() {
+
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: can delete empty line

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+            return count;
+        }
+
+        @Override
+        public byte[] position() {
+            return this.position;
+        }
+
+        @Override
+        public void close() {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:558]: need to close this.iter?

File: src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2019 HugeGraph Authors
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java:2]: expect apache license

File: src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    public String supportsMinVersion() {
+        return "0.11.2";
+    }
+
+    @Override
+    public String supportsMaxVersion() {
+        return "0.12";
+    }
+
+    public static void main(String[] args) {
```
Comment (zyxxoo) [inline src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java:39]: Is this code for debugging?  maybe should del

File: src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+public class TikvPlugin implements HugeGraphPlugin {
+
+    @Override
+    public String name() {
+        return TikvStoreProvider.TYPE;
+    }
+
+    @Override
+    public void register() {
```
Comment (zyxxoo) [inline src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java:20]: is there should add todo for registering serializer and options?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStore.java [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+        if (this.sessions == null) {
+            this.sessions = new TikvStdSessions(config, this.namespace, this.store);
+        }
+
+        assert this.sessions != null;
+        if (!this.sessions.closed()) {
+            LOG.debug("Store {} has been opened before", this.store);
+            this.sessions.useSession();
+            return;
```
Comment (zyxxoo) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStore.java:138]: else if (!this.session.closed()) {
 ....
return;
}

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+        @Override
+        public void deleteRange(String table, byte[] keyFrom, byte[] keyTo) {
+            ByteString startKey = this.key(table, keyFrom);
+            ByteString endKey = this.key(table, keyTo);
+            this.deleteRangeBatch.put(startKey, endKey);
+        }
+
+        @Override
+        public byte[] get(String table, byte[] key) {
+            return tikv().get(this.key(table, key)).toByteArray();
```
Comment (zyxxoo) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:305]: should we merge cache data (putBatch, delBatch...)?

File: src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        return "0.11.2";
+    }
+
+    @Override
+    public String supportsMaxVersion() {
+        return "0.12";
+    }
+
+    public static void main(String[] args) {
+        new com.baidu.hugegraph.tikv.TikvPlugin().register();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java]: can be `new TikvPlugin().register()`

File: src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java (hunk 9 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/tikv/TikvPlugin.java:20]: yes, expect to call registerOptions() here

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvSessions.java (hunk 3 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvSessions.java]: not addressed: improve comments related to "RocksDB"

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 6 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:558]: not addressed: need to close this.iter?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 11 above)
Comment (zhoney) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:305]: no need. query will handle none committed data in upper layer.

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 6 above)
Comment (zhoney) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:558]: this.iter is just a regular Iterator(none closable)

**Review by javeme — APPROVED** [review]

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 11 above)
Comment (zyxxoo) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:305]: is there 'assert !this.hasChanges()' is missed?

**Pull request merged**